src/Database/LSMTree/Internal/TableHandle.hs

```diff
@@ -0,0 +1,191 @@
+    LSMTreeError (..)
+    -- * Session
+  , Session (..)
+    -- * Table handle
+  , TableHandle
+    -- * configuration
+  , TableConfig (..)
+  , ResolveMupsert (..)
+  , SizeRatio (..)
+  , MergePolicy (..)
+  , BloomFilterAlloc (..)
+  ) where
+
+import           Control.Concurrent.Class.MonadMVar.Strict (StrictMVar)
+import           Control.Exception (Exception)
+import           Data.BloomFilter (Bloom)
+import qualified Data.Vector as V
+import           Data.Word (Word32)
+import           Database.LSMTree.Internal.IndexCompact (IndexCompact)
+import           Database.LSMTree.Internal.Run (Run)
+import qualified Database.LSMTree.Internal.Run as Run
+import           Database.LSMTree.Internal.Serialise (SerialisedKey,
+                     SerialisedValue)
+import           Database.LSMTree.Internal.WriteBuffer (WriteBuffer)
+import           System.FS.API (FsErrorPath, FsPath, HasFS)
+import           System.FS.BlockIO.API (HasBlockIO)
+
+{-------------------------------------------------------------------------------
+  Exceptions
+-------------------------------------------------------------------------------}
+
+-- TODO: give this a nicer Show instance.
+data LSMTreeError =
+    SessionDirDoesNotExist FsErrorPath
+  | SessionDirLocked FsErrorPath
+  deriving (Show, Exception)
+
+{-------------------------------------------------------------------------------
+  Session
+-------------------------------------------------------------------------------}
+
+-- | A session provides context that is shared across multiple table handles.
+--
+-- For more information, see 'Database.LSMTree.Common.Session'.
+data Session m h = Session {
+    -- | The path to the directory in which this sesion is live. This is a path
+    -- relative to root of the 'HasFS' instance.
+    sessionRoot        :: !FsPath
+  , sessionHasFS       :: !(HasFS m h)
+  , sessionHasBlockIO  :: !(HasBlockIO m h)
+    -- | A session-wide shared, atomic counter that is used to produce unique
+    -- names for run files.
+  , sessionUniqCounter :: !(StrictMVar m Integer)
+  }
+
+{-------------------------------------------------------------------------------
+  Table handle
+-------------------------------------------------------------------------------}
+
+-- | A handle to an on-disk key\/value table.
+--
+-- For more information, see 'Database.LSMTree.Normal.TableHandle'.
+--
+-- NOTE: a table handle is __not__ thread-safe.
+data TableHandle m h = TableHandle {
+    tableSession     :: !(Session m h)
+  , tableConfig      :: !TableConfig
+  , tableWriteBuffer :: !(StrictMVar m (WriteBuffer))
+    -- | A hierarchy of levels. The vector indexes double as level numbers.
+  , tableLevels      :: !(StrictMVar m (V.Vector (Level h)))
+    -- | Cache of flattened 'levels'.
+    --
+    -- INVARIANT: when 'level's is modified, this cache should be updated as
+    -- well, for example using 'mkLevelsCache'.
+  , tableLevelsCache :: !(StrictMVar m (LevelsCache h))
```

Comment:
We might want to iterate on the exact synchronisation mechanism (e.g. separate `MVar`s for the levels and levels cache seems like a risk), but I think it's okay for now and should not hold up this PR.

Reply:
Let's do that